File: leetcode/95. Unique Binary Search Trees II/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func helper(s, n int) []*TreeNode {
	ans := []*TreeNode{}

	if s > n {
		return []*TreeNode{nil}
	}

	for i := s; i <= n; i++ {
		for _, left := range helper(s, i-1) {
			for _, right := range helper(i+1, n) {
				ans = append(ans, &TreeNode{
					Val:   i,
					Left:  left,
					Right: right,
				})
			}
		}
	}

	return ans
}

func generateTrees(n int) []*TreeNode {
	return helper(1, n)
}

File: leetcode/95. Unique Binary Search Trees II/main.java
import java.util.*;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<TreeNode> generateTrees(int n) {
        return this.helper(1, n);
    }

    private List<TreeNode> helper(int s, int n) {
        List<TreeNode> ans = new LinkedList<>();

        if (s > n) {
            ans.add(null);
            return ans;
        }

        for (int i = s; i <= n; i++) {
            for (TreeNode left : this.helper(s, i - 1)) {
                for (TreeNode right : this.helper(i + 1, n)) {
                    ans.add(new TreeNode(i, left, right));
                }
            }
        }

        return ans;
    }
}